### PR TITLE
RUMM-440 Condition the OkHttp APM tracing on provided domains list

### PR DESCRIPTION
**Pull request by mariusc83**
### Review checklist (to be filled by reviewers)

- [x] Feature or bugfix MUST have appropriate tests (unit, integration, e2e)
- [x] Make sure you discussed the feature or bugfix with the maintaining team in an Issue
- [x] Make sure each commit and the PR mention the Issue number (cf the [CONTRIBUTING](CONTRIBUTING.md) doc)

### CODE REVIEW

File: dd-sdk-android/src/test/kotlin/com/datadog/android/tracing/internal/net/TracingRequestInterceptorTest.kt [hunk 1]
```diff
@@ -133,22 +134,30 @@ internal class TracingRequestInterceptorTest {
     lateinit var fakePackageVersion: String
 
     @BeforeEach
-    fun `set up`() {
+    fun `set up`(forge: Forge) {
+        fakeWhitelistedDomain = forge.aStringMatching("[a-z0-9_]{8}\\.[a-z]{3}")
```
Comment (xgouchet) [inline dd-sdk-android/src/test/kotlin/com/datadog/android/tracing/internal/net/TracingRequestInterceptorTest.kt:138]: The [RFC 3986](https://tools.ietf.org/html/rfc3986) gives the following definition of what a host name can be
```
reg-name      = *( unreserved / pct-encoded / sub-delims )
unreserved    = ALPHA / DIGIT / "-" / "." / "_" / "~"
sub-delims    = "!" / "$" / "&" / "'" / "(" / ")" / "*" / "+" / "," / ";" / "="
pct-encoded   = "%" HEXDIG HEXDIG // e.g.: %20
```

You're only testing a trivial subset, it'd be nice to be sure the code works for any valid host. Especially since you're generating a regex that could embed special characters.

File: dd-sdk-android/src/main/kotlin/com/datadog/android/tracing/TracingInterceptor.kt [hunk 2]
```diff
@@ -26,11 +26,14 @@ import okhttp3.OkHttpClient
...
  *   OkHttpClient client = new OkHttpClient.Builder()
- *       .addInterceptor(new TracingInterceptor())
+ *       .addInterceptor(new TracingInterceptor(listOf("yourdomain.com")))
  *       // Optional RUM integration
  *       .addInterceptor(new RumInterceptor())
  *       .build();
  * ```
+ * @param apmTrackedDomains a list of all the domains that you want to be automatically tracked
+ * by this interceptor. Please note that if no domain is provided in the list our
+ * tracing interceptor will not automatically track and create any Span for your OkHttp requests.
```
Comment (xgouchet) [inline dd-sdk-android/src/main/kotlin/com/datadog/android/tracing/TracingInterceptor.kt:36]: Same as above ;)

File: dd-sdk-android/src/main/kotlin/com/datadog/android/tracing/internal/net/TracingRequestInterceptor.kt [hunk 3]
```diff
@@ -25,7 +25,25 @@ import java.util.concurrent.atomic.AtomicReference
...
+        val regexFormat =
+            if (acceptedDomains.isEmpty()) {
+                ""
+            } else {
+                acceptedDomains
+                    .map { "^$it" }
+                    .reduce { current, next ->
+                        "$current|$next"
+                    }
+                    .plus("(.*)")
```
[the remark this answers is not in the document]
Comment (mariusc83) [reply dd-sdk-android/src/main/kotlin/com/datadog/android/tracing/internal/net/TracingRequestInterceptor.kt:43]: You are right about the false positives ....I actually wanted to remove that from there....don't know why I left it.

File: dd-sdk-android/src/test/kotlin/com/datadog/android/tracing/internal/net/TracingRequestInterceptorTest.kt [hunk 4]
```diff
@@ -133,22 +134,30 @@ internal class TracingRequestInterceptorTest {
...
+        fakeSecondWhitelistedDomain = forge.aStringMatching("[a-z0-9_]{8}\\.[a-z]{3}")
+        val fakeBlacklistedDomain = forge.anAlphabeticalString(size = 1) + fakeWhitelistedDomain
+        fakeWhitelistedDomainUrl = "http://$fakeWhitelistedDomain/"
+        fakeWhitelistedSubdomainUrl = "http://$fakeWhitelistedDomain." +
```
[the remark this answers is not in the document]
Comment (mariusc83) [reply dd-sdk-android/src/test/kotlin/com/datadog/android/tracing/internal/net/TracingRequestInterceptorTest.kt:142]: My bad about subdomains...was actually thinking at rest methods :( when I did that example...will add more tests in there anyway.

File: .gitlab-ci.yml [hunk 5]
```diff
@@ -1,4 +1,4 @@
-include:
+ include:
```
Comment (xgouchet) [inline .gitlab-ci.yml:1]: Why did you change this ? YAML is very sensitive to indentation changes, not sure this is what you want.

File: dd-sdk-android/src/test/kotlin/com/datadog/android/tracing/internal/net/TracingRequestInterceptorTest.kt [hunk 6]
```diff
@@ -189,7 +192,7 @@ internal class TracingRequestInterceptorTest {
         assertThat(transformedRequest.method()).isEqualTo(fakeMethod)
         assertThat(transformedRequest.header(key)).isEqualTo(value)
         inOrder(mockSpan) {
-            verify(mockSpan).setTag("http.url", fakeUrl)
+            verify(mockSpan).setTag("http.url", fakeWhitelistedDomainUrl.toLowerCase())
```
Comment (xgouchet) [inline dd-sdk-android/src/test/kotlin/com/datadog/android/tracing/internal/net/TracingRequestInterceptorTest.kt:195]: Why the "toLowerCase()"? Urls are case sensitive for everything after the domain name, changing the case means it's not the same URL. You should instead generate a `fakeWhitelistedDomainUrl` based on the lowercase of `fakeWhitelistedDomain`

File: .gitlab-ci.yml (hunk 5 above)
Comment (mariusc83) [reply .gitlab-ci.yml:1]: probably was ktlint :(

File: dd-sdk-android/src/test/kotlin/com/datadog/android/tracing/internal/net/TracingRequestInterceptorTest.kt (hunk 6 above)
Comment (mariusc83) [reply dd-sdk-android/src/test/kotlin/com/datadog/android/tracing/internal/net/TracingRequestInterceptorTest.kt:195]: the fake urls are only "http://domain/" now. The `request.url().host()` is actually transforming the host to lowercase so I chose the easy way to assert those as will not affect any test...but yeah it is a bit ugly. Need to rework also those url generators.

Comment (mariusc83) [reply dd-sdk-android/src/test/kotlin/com/datadog/android/tracing/internal/net/TracingRequestInterceptorTest.kt:195]: Because we also generate domains with capital letters but OkHttp transforms the domain name to lowercase when you call `request.url()`.  Maybe I will create a method that computes the *expectedUrl* based on the *fakeUrl* by lowering the case of the *domain* in it.
